drivers/cartesi/cmio.c

```diff
@@ -0,0 +1,231 @@
+	if (!cmio)
+		return -EBADF;
+
+	switch (cmd) {
+	case IOCTL_CMIO_SETUP:
+		return cmio_ioctl_setup(cmio, arg);
+	case IOCTL_CMIO_YIELD:
+		return cmio_ioctl_yield(cmio, arg);
+	}
+	return -ENOIOCTLCMD;
+}
+
+static int cmio_mmap(struct file *file, struct vm_area_struct *vma)
+{
+	__u64 vma_size;
+	struct cmio_device *cmio = to_cmio_device(file);
+	if (!cmio)
+		return -EBADF;
+
+	vma_size = vma->vm_end - vma->vm_start;
+	if (!((vma->vm_start == cmio->bufs.tx.data && vma_size == cmio->bufs.tx.length)
+	||    (vma->vm_start == cmio->bufs.rx.data && vma_size == cmio->bufs.rx.length)))
+		return -EINVAL;
+
+	return remap_pfn_range(vma,
+	                       vma->vm_start,
+	                       vma->vm_start >> PAGE_SHIFT,
+	                       vma_size,
+	                       vma->vm_page_prot);
+}
+
+static const struct file_operations cmio_fileops = {
+	.open           = cmio_open,
+	.release        = cmio_release,
+	.unlocked_ioctl = cmio_ioctl,
+	.mmap           = cmio_mmap,
+	.owner          = THIS_MODULE,
+};
+
+static int setup_buffer(struct device_node *parent
+                       ,const char *name
+                       ,struct cmio_buffer *buf)
+{
+	u64 xs[2];
+	int rc = -EIO;
+	struct device_node *node = NULL;
+
+	if (!(node = of_find_node_by_name(parent, name)))
+		goto leave;
+	if (of_property_read_u64_array(node, "reg", xs, 2))
+		goto leave;
+	if (!(buf->data = xs[0]))
+		goto leave;
+	if (!(buf->length = xs[1]))
+		goto leave;
+	rc = 0;
+leave:
+	of_node_put(node);
+	return rc;
+}
+
+static int setup_io(struct cmio_device *cmio)
+{
+	int rc = -EIO;
+	struct device_node *cmio_node = NULL;
+
+	if (!(cmio_node = of_find_node_by_path("/rollup"))
+	||  setup_buffer(cmio_node, "tx_buffer", &cmio->bufs.tx)
+	||  setup_buffer(cmio_node, "rx_buffer", &cmio->bufs.rx))
+		goto leave;
+	rc = 0;
+leave:
+	of_node_put(cmio_node);
+	return rc;
+}
+
+static int cmio_driver_probe(struct platform_device *pdev)
+{
+	int rc;
+	struct cmio_device *cmio;
+
+	cmio = devm_kzalloc(&pdev->dev, sizeof(*cmio), GFP_KERNEL);
+	if (!cmio)
+		return -ENOMEM;
+
+	atomic_set(&cmio->single_user_lock, 1);
+	cmio->mdev.minor = MISC_DYNAMIC_MINOR;
+	cmio->mdev.name  = DEVICE_NAME;
+	cmio->mdev.fops  = &cmio_fileops;
+	rc = misc_register(&cmio->mdev);
+	if (rc) {
+		dev_err(&pdev->dev, "failed to register miscdevice\n");
+		goto leave;
+	}
+
+	rc = setup_io(cmio);
+	if (rc) {
+		dev_err(&pdev->dev, "failed to parse device tree\n");
+		goto deregister;
+	}
+
+	platform_set_drvdata(pdev, cmio);
+	cmio->pdev = pdev;
+
+	pr_info(MODULE_DESC ": Module loaded\n");
+	return 0;
+
+deregister:
+	misc_deregister(&cmio->mdev);
+leave:
+	return rc;
+}
+
+static int cmio_driver_remove(struct platform_device *pdev)
+{
+	struct cmio_device *cmio = platform_get_drvdata(pdev);
+	misc_deregister(&cmio->mdev);
+	dev_info(&pdev->dev, "unregistered\n");
+	return 0;
+}
+
+static const struct of_device_id cmio_match[] = {
+	{.compatible = "ctsi-rollup",}, {},
```

Comment:
Change this to `"ctsi-cmio"`? Do all our things have `ctsi-` before?

Reply:
> Change this to "ctsi-cmio"?

ok

> Do all our things have `ctsi-` before?

As far back as I remember.